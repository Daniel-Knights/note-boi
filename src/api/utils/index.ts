export * from './decorators';
export * from './error';
export * from './http';
