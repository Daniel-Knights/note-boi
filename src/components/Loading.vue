<template>
  <div id="loading">
    <svg class="loading__spinner" viewBox="0 0 95 95">
      <g transform="translate(-0.88095856,-1.2589302)">
        <circle cx="48.380959" cy="48.75893" r="40" />
      </g>
    </svg>
  </div>
</template>

<script lang="ts" setup></script>

<style lang="scss">
#loading {
  @include v.flex-x(center, center);
  position: fixed;
  @include v.cover;
  background-color: var(--colour__primary);
  z-index: 50;
}

.loading__spinner {
  display: block;
  padding: 10px 0;
  @include v.equal-dimensions(120px);
  animation: rotate 1s ease-out infinite;

  circle {
    fill-opacity: 0;
    stroke: var(--colour__secondary);
    stroke-width: 2;
    stroke-dasharray: 262;
    stroke-linecap: round;
    animation: loading-spinner 1s linear infinite;

    @keyframes loading-spinner {
      0%,
      100% {
        stroke-dashoffset: 72;
      }
      50% {
        stroke-dashoffset: 240;
      }
    }
  }

  @keyframes rotate {
    from {
      transform: rotate(-25deg);
    }
    to {
      transform: rotate(335deg);
    }
  }
}
</style>
