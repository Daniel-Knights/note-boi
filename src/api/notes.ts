import {
  AppError,
  DebounceQueue,
  Encryptor,
  ERROR_CODE,
  ErrorConfig,
  FetchBuilder,
  KeyStore,
} from '../classes';
import { DecryptedNoteDiff, NOTE_EVENTS } from '../constant';
import {
  findNote,
  newNote,
  noteState,
  selectNote,
  sortStateNotes,
  UnsyncedEventDetail,
} from '../store/note';
import { resetAppError, syncState } from '../store/sync';
import { isEmptyNote, tauriEmit, tauriInvoke } from '../utils';

import {
  parseErrorRes,
  resIsOk,
  route,
  throwEncryptorError,
  throwFetchError,
} from './utils';

const syncQueue = new DebounceQueue();

// Sync
export const sync = route(async (isCancelled?: () => boolean) => {
  const errorConfig = {
    code: ERROR_CODE.SYNC,
    retry: { fn: sync, args: [isCancelled] },
    display: { sync: true },
  } satisfies Omit<ErrorConfig<typeof sync>, 'message'>;

  const passwordKey = await KeyStore.getKey();
  if (!passwordKey || isCancelled?.()) return;

  const notesToEncrypt = noteState.notes.filter((nt) => {
    const noteIsEdited = syncState.unsyncedNoteIds.edited.has(nt.id);
    const noteIsCached = syncState.encryptedNotesCache.has(nt.id);

    return (noteIsEdited || !noteIsCached) && !isEmptyNote(nt);
  });

  const [accessToken, encryptedNotes] = await Promise.all([
    tauriInvoke('get_access_token', {
      username: syncState.username,
    }),
    Encryptor.encryptNotes(notesToEncrypt, passwordKey).catch((err) =>
      throwEncryptorError(errorConfig, err)
    ),
  ]);
  if (!accessToken || !encryptedNotes || isCancelled?.()) return;

  encryptedNotes.forEach((nt) => {
    syncState.encryptedNotesCache.set(nt.id, nt);
  });

  const res = await new FetchBuilder('/notes/sync')
    .method('PUT')
    .withAuth(syncState.username, accessToken)
    .body({
      notes: [...syncState.encryptedNotesCache.values()],
      deleted_note_ids: [...syncState.unsyncedNoteIds.deleted],
    })
    .fetch(syncState.username)
    .catch((err) => throwFetchError(errorConfig, err));
  if (!res || isCancelled?.()) return;

  if (resIsOk(res)) {
    resetAppError();
    tauriEmit('auth', { is_logged_in: true });

    // Users' session must still be valid
    syncState.isLoggedIn = true;

    const decryptedNotes = await Promise.all([
      Encryptor.decryptNotes(res.data.note_diff.added, passwordKey),
      Encryptor.decryptNotes(res.data.note_diff.edited, passwordKey),
    ]).catch((err) => throwEncryptorError(errorConfig, err));
    if (!decryptedNotes || isCancelled?.()) return;

    await updateLocalNoteStateFromDiff({
      added: decryptedNotes[0],
      edited: decryptedNotes[1],
      deleted_ids: res.data.note_diff.deleted_ids,
    });
  } else {
    throw new AppError({
      ...errorConfig,
      message: parseErrorRes(res),
    });
  }
});

/**
 * Updates local note state based on the given diff.
 */
export function updateLocalNoteStateFromDiff(noteDiff: DecryptedNoteDiff) {
  let selectedNoteIsDeleted = false;
  let remoteSelectedNote;

  // Resolve edited and deleted note conflicts
  for (let i = noteState.notes.length - 1; i >= 0; i -= 1) {
    const ln = noteState.notes[i]!;
    const isSelectedNote = ln.id === noteState.selectedNote.id;

    // Remove notes that were deleted on the server
    if (noteDiff.deleted_ids.includes(ln.id)) {
      noteState.notes.splice(i, 1);

      if (isSelectedNote) {
        selectedNoteIsDeleted = true;
      }

      continue;
    }

    // Update existing notes
    const foundRemoteNoteIndex = noteDiff.edited.findIndex((rn) => rn.id === ln.id);

    if (foundRemoteNoteIndex > -1) {
      noteState.notes[i] = noteDiff.edited[foundRemoteNoteIndex]!;

      noteDiff.edited.splice(foundRemoteNoteIndex, 1);

      if (isSelectedNote) {
        remoteSelectedNote = noteState.notes[i];
      }
    }
  }

  // Add new notes from the server
  noteDiff.added.forEach((rn) => {
    noteState.notes.push(rn);
  });

  sortStateNotes();

  // New note if no notes exist
  if (noteState.notes.length === 0) {
    newNote();
  }
  // Select next note if current selected note was deleted
  else if (selectedNoteIsDeleted) {
    selectNote(noteState.notes[0]!.id);
  }
  // Update selected note if edited and ensure editor updates
  else if (remoteSelectedNote) {
    noteState.selectedNote.content = remoteSelectedNote.content;
    noteState.selectedNote.timestamp = remoteSelectedNote.timestamp;

    document.dispatchEvent(new Event(NOTE_EVENTS.change));
  }
  // Select next note if current selected note is empty and not deliberately created
  else if (noteState.notes.length > 1 && !syncState.unsyncedNoteIds.new) {
    const foundNote = findNote(noteState.selectedNote.id);

    if (!foundNote || isEmptyNote(foundNote)) {
      selectNote(noteState.notes[1]!.id);
    }
  }

  syncState.unsyncedNoteIds.clear();

  return tauriInvoke('sync_local_notes', { notes: noteState.notes });
}

/**
 * Adds sync call to the debounce queue.
 * `isInitial` indicates this is the initial sync on app load.
 */
export function debounceSync(isInitial = false): void {
  if (!isInitial && !syncState.isLoggedIn) return;

  syncQueue.add(
    (isCancelled) => {
      return sync(isCancelled);
    },
    isInitial ? undefined : 500
  );
}

// Keep track of notes with unsynced changes
document.addEventListener(
  NOTE_EVENTS.unsynced,
  (ev: CustomEventInit<UnsyncedEventDetail>) => {
    if (!ev.detail) return;

    syncState.unsyncedNoteIds.set({ [ev.detail.kind]: [ev.detail.noteId] });
  }
);
