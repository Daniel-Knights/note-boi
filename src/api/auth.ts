import {
  AppError,
  Encryptor,
  ERROR_CODE,
  ErrorConfig,
  FetchBuilder,
  KeyStore,
  Storage,
} from '../classes';
import { noteState } from '../store/note';
import { resetAppError, syncState } from '../store/sync';
import { isEmptyNote, tauriEmit, tauriInvoke } from '../utils';

import { updateLocalNoteStateFromDiff } from './notes';
import {
  parseErrorRes,
  resIsOk,
  route,
  throwEncryptorError,
  throwFetchError,
} from './utils';

export function clientSideLogout(): Promise<void> {
  tauriInvoke('delete_access_token', { username: syncState.username });

  syncState.username = '';
  syncState.isLoggedIn = false;

  Storage.remove('USERNAME');
  KeyStore.reset();

  return tauriEmit('auth', { is_logged_in: false });
}

// Login
export const login = route(async () => {
  const errorConfig = {
    code: ERROR_CODE.LOGIN,
    retry: { fn: login },
    display: {
      form: true,
      sync: true,
    },
  } satisfies Omit<ErrorConfig<typeof login>, 'message'>;

  const passwordKey = await Encryptor.generatePasswordKey(syncState.password);

  const encryptedNotes = await Encryptor.encryptNotes(
    noteState.notes.filter((nt) => !isEmptyNote(nt)),
    passwordKey
  ).catch((err) => throwEncryptorError(errorConfig, err));
  if (!encryptedNotes) return;

  encryptedNotes.forEach((nt) => {
    syncState.encryptedNotesCache.set(nt.id, nt);
  });

  const res = await new FetchBuilder('/auth/login')
    .method('POST')
    .body({
      username: syncState.username,
      password: syncState.password,
      notes: encryptedNotes,
      deleted_notes: syncState.unsyncedNotes.deleted,
    })
    .fetch(syncState.username)
    .catch((err) => throwFetchError(errorConfig, err));
  if (!res) return;

  if (resIsOk(res)) {
    syncState.password = '';
    syncState.isLoggedIn = true;

    Storage.set('USERNAME', syncState.username);

    resetAppError();
    tauriEmit('auth', { is_logged_in: true });

    await KeyStore.storeKey(passwordKey);

    const decryptedNotes = await Promise.all([
      Encryptor.decryptNotes(res.data.note_diff.added, passwordKey),
      Encryptor.decryptNotes(res.data.note_diff.edited, passwordKey),
    ]).catch((err) => throwEncryptorError(errorConfig, err));
    if (!decryptedNotes) return;

    await updateLocalNoteStateFromDiff({
      added: decryptedNotes[0],
      edited: decryptedNotes[1],
      deleted: res.data.note_diff.deleted,
    });
  } else {
    throw new AppError({
      ...errorConfig,
      message: parseErrorRes(res),
    });
  }
});

// Signup
export const signup = route(async () => {
  const errorConfig = {
    code: ERROR_CODE.SIGNUP,
    retry: { fn: signup },
    display: {
      form: true,
      sync: true,
    },
  } satisfies Omit<ErrorConfig<typeof signup>, 'message'>;

  const passwordKey = await Encryptor.generatePasswordKey(syncState.password);

  await KeyStore.storeKey(passwordKey);

  const encryptedNotes = await Encryptor.encryptNotes(
    noteState.notes.filter((nt) => !isEmptyNote(nt)),
    passwordKey
  ).catch((err) => throwEncryptorError(errorConfig, err));
  if (!encryptedNotes) return;

  const res = await new FetchBuilder('/auth/signup')
    .method('POST')
    .body({
      username: syncState.username,
      password: syncState.password,
      notes: encryptedNotes,
    })
    .fetch(syncState.username)
    .catch((err) => throwFetchError(errorConfig, err));
  if (!res) return;

  if (resIsOk(res)) {
    resetAppError();
    tauriEmit('auth', { is_logged_in: true });

    syncState.password = '';
    syncState.isLoggedIn = true;
    syncState.unsyncedNotes.clear();

    Storage.set('USERNAME', syncState.username);
  } else {
    throw new AppError({
      ...errorConfig,
      message: parseErrorRes(res),
    });
  }
});

// Logout
export const logout = route(async () => {
  const errorConfig = {
    code: ERROR_CODE.LOGOUT,
    retry: { fn: logout },
    display: { form: true },
  } satisfies Omit<ErrorConfig<typeof logout>, 'message'>;

  const accessToken = await tauriInvoke('get_access_token', {
    username: syncState.username,
  });

  const fetchPromise = new FetchBuilder('/auth/logout')
    .method('POST')
    .withAuth(syncState.username, accessToken)
    .fetch()
    .catch((err) => throwFetchError(errorConfig, err));

  const [res] = await Promise.all([fetchPromise, clientSideLogout()]);
  if (!res) return;

  if (resIsOk(res)) {
    resetAppError();
  } else {
    // No use bringing this to the user's attention
    throw new AppError({
      code: ERROR_CODE.LOGOUT,
      message: parseErrorRes(res),
    });
  }
});
