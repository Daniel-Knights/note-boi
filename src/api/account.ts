import * as dialog from '@tauri-apps/plugin-dialog';

import {
  AppError,
  Encryptor,
  ERROR_CODE,
  ErrorConfig,
  FetchBuilder,
  KeyStore,
} from '../classes';
import { noteState } from '../store/note';
import { resetAppError, syncState } from '../store/sync';
import { tauriInvoke } from '../utils';

import { clientSideLogout } from './auth';
import {
  parseErrorRes,
  resIsOk,
  route,
  throwEncryptorError,
  throwFetchError,
} from './utils';

export const changePassword = route(async (): Promise<void> => {
  const errorConfig = {
    code: ERROR_CODE.CHANGE_PASSWORD,
    retry: { fn: changePassword },
    display: {
      form: true,
      sync: true,
    },
  } satisfies Omit<ErrorConfig<typeof changePassword>, 'message'>;

  const newKey = await Encryptor.generatePasswordKey(syncState.newPassword);

  await KeyStore.storeKey(newKey);

  const [accessToken, encryptedNotes] = await Promise.all([
    tauriInvoke('get_access_token', {
      username: syncState.username,
    }),
    Encryptor.encryptNotes(noteState.notes, newKey).catch((err) =>
      throwEncryptorError(errorConfig, err)
    ),
  ]);
  if (!accessToken || !encryptedNotes) return;

  const res = await new FetchBuilder('/account/change-password')
    .method('PUT')
    .withAuth(syncState.username, accessToken)
    .body({
      current_password: syncState.password,
      new_password: syncState.newPassword,
      notes: encryptedNotes,
    })
    .fetch(syncState.username)
    .catch((err) => throwFetchError(errorConfig, err));
  if (!res) return;

  if (resIsOk(res)) {
    resetAppError();

    syncState.password = '';
    syncState.newPassword = '';
  } else {
    throw new AppError({
      ...errorConfig,
      message: parseErrorRes(res),
    });
  }
});

export const deleteAccount = route(async (): Promise<void> => {
  const askRes = await dialog.ask('Are you sure?', {
    title: 'Delete account',
    kind: 'warning',
  });
  if (!askRes) return;

  const errorConfig = {
    code: ERROR_CODE.DELETE_ACCOUNT,
    retry: { fn: deleteAccount },
    display: {
      sync: true,
    },
  } satisfies Omit<ErrorConfig<typeof deleteAccount>, 'message'>;

  const accessToken = await tauriInvoke('get_access_token', {
    username: syncState.username,
  });

  const res = await new FetchBuilder('/account/delete')
    .method('DELETE')
    .withAuth(syncState.username, accessToken)
    .fetch()
    .catch((err) => throwFetchError(errorConfig, err));
  if (!res) return;

  if (resIsOk(res)) {
    await clientSideLogout();
    resetAppError();
    syncState.unsyncedNotes.clear(true);
  } else {
    throw new AppError({
      ...errorConfig,
      message: parseErrorRes(res),
    });
  }
});
