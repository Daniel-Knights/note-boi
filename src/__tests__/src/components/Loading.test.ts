import { mount } from '@vue/test-utils';

import { mockApi } from '../../api';

import Loading from '../../../components/Loading.vue';

describe('Loading', () => {
  it('Mounts', async () => {
    const { calls, promises } = mockApi();
    const wrapper = mount(Loading);

    await Promise.all(promises);

    assert.isTrue(wrapper.isVisible());
    assert.strictEqual(calls.size, 0);
  });
});
