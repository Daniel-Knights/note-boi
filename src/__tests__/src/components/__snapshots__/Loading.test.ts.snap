// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Loading > Mounts > KeyStore.getKey() 1`] = `undefined`;

exports[`Loading > Mounts > allCalls 1`] = `[]`;

exports[`Loading > Mounts > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`Loading > Mounts > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`Loading > Mounts > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`Loading > Mounts > openedPopup 1`] = `undefined`;

exports[`Loading > Mounts > selectedTheme 1`] = `"System"`;

exports[`Loading > Mounts > storage.get('MENU_WIDTH') 1`] = `null`;

exports[`Loading > Mounts > storage.get('THEME') 1`] = `null`;

exports[`Loading > Mounts > storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`Loading > Mounts > storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`Loading > Mounts > storage.get('USERNAME') 1`] = `null`;

exports[`Loading > Mounts > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoading": false,
  "isLoggedIn": false,
  "isLogin": true,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "add": [Function],
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "size": 0,
  },
  "username": "",
}
`;

exports[`Loading > Mounts > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;
